### PR TITLE
sd-event: re-check new epoll events when a child event is queued

### LINKED ISSUE #18190: The program runs successfully and has output, but systemd-udevd loses program's output information and the stdout event appears to be unprocessed.
**systemd version the issue has been seen with**
 > v247

**Used distribution**
 > f33

**Linux kernel version used** (`uname -a`)
 > 5.9.16

**CPU architecture issue was seen on**
 > aarch64

**Expected behaviour you didn't see**
 > Everything goes well.

**Unexpected behaviour you saw**
 > The program runs successfully and has output, but systemd-udevd loses program's output information and the stdout event appears to be unprocessed.

**Steps to reproduce the problem**
 > 
```
#!/bin/bash

function check_lvm(){
    for i in `seq 10`;do
        vgs | grep vgtest$i || exit 1
        ls /dev/vgtest$i || exit 1
        for j in `seq 10`;do
            lvs | grep lvtest${i}-${j} || exit 1
            ls /dev/vgtest${i}/lvtest${i}-${j} || exit 1
        done
    done
}

while true;do
    vgchange --refresh
    check_lvm
```
/usr/lib/udev/rules.d/11-dm-lvm.rules
```
# Use DM name and split it up into its VG/LV/layer constituents.
IMPORT{program}="/usr/sbin/dmsetup splitname --nameprefixes --noheadings --rows $env{DM_NAME}"    ####HERE, program cannot got the output ####

# DM_SUBSYSTEM_UDEV_FLAG0 is the 'NOSCAN' flag for LVM subsystem.
# This flag is used to temporarily disable selected rules to prevent any
# processing or scanning done on the LVM volume before LVM has any chance
# to zero any stale metadata found within the LV data area. Such stale
# metadata could cause false claim of the LV device, keeping it open etc.
#
# If the NOSCAN flag is present, backup selected existing flags used to
# disable rules, then set them firmly so those selected rules are surely skipped.
# Restore these flags once the NOSCAN flag is dropped (which is normally any
# uevent that follows for this LV, even an artificially generated one).
ENV{DM_SUBSYSTEM_UDEV_FLAG0}=="1", ENV{DM_NOSCAN}="1", ENV{DM_DISABLE_OTHER_RULES_FLAG_OLD}="$env{DM_UDEV_DISABLE_OTHER_RULES_FLAG}", ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}="1"
ENV{DM_SUBSYSTEM_UDEV_FLAG0}!="1", IMPORT{db}="DM_NOSCAN", IMPORT{db}="DM_DISABLE_OTHER_RULES_FLAG_OLD"
ENV{DM_SUBSYSTEM_UDEV_FLAG0}!="1", ENV{DM_NOSCAN}=="1", ENV{DM_UDEV_DISABLE_OTHER_RULES_FLAG}="$env{DM_DISABLE_OTHER_RULES_FLAG_OLD}", \
                                   ENV{DM_DISABLE_OTHER_RULES_FLAG_OLD}="", ENV{DM_NOSCAN}=""

ENV{DM_UDEV_DISABLE_SUBSYSTEM_RULES_FLAG}=="1", GOTO="lvm_end"

# Do not create symlinks for inappropriate subdevices.
ENV{DM_LV_NAME}=="pvmove?*|?*_vorigin", GOTO="lvm_disable"
ENV{DM_LV_LAYER}=="?*", GOTO="lvm_disable"

# Create symlinks for top-level devices only.
ENV{DM_VG_NAME}=="?*", ENV{DM_LV_NAME}=="?*", SYMLINK+="$env{DM_VG_NAME}/$env{DM_LV_NAME}", GOTO="lvm_end"   ###The SYMLINK cannot be created correctly.###
```
The link cannot be created correctly because udev does not get output, but we are sure that the dmsetup has output.

### PR DESCRIPTION
**Pull request by yuwata**
Previously, when a process outputs something and exit just after
epoll_wait() but before process_child(), then the IO event is ignored
even if the IO event has higher priority. See #18190.

This can be solved by checking epoll event again after process_child().

However, there exists a possibility that another process outputs and
exits just after process_child() but before the second epoll_wait().
When the IO event has lower priority than the child event, still IO
event is processed.

So, this makes new epoll events and child events are checked in a loop
until no new event is detected. To prevent an infinite loop, the number
of maximum trial is set to 10.

Fixes #18190.

Replaces #18191, #18750, and #18799.

### CODE REVIEW

**Comment by yuwata** [comment]
Yey, all CentOS CIs are green now, previously they were all red in #18922. PTAL.

**Comment by yuwata** [comment]
Yey, all green!

File: src/libsystemd/sd-event/sd-event.c [hunk 1]
```diff
@@ -3920,19 +3938,75 @@ _public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
...
+                if (r < 0)
+                        goto finish;
+                if (r == 0 && threshold < INT64_MAX)
+                        /* No new epoll event. */
+                        break;
+
+                r = process_child(e, threshold, &child_min_priority);
                 if (r < 0)
                         goto finish;
+                if (r == 0)
```
Comment (bluca) [inline src/libsystemd/sd-event/sd-event.c:4001]: this was `if (r == 0 && threshold < INT64_MAX)` in the previous patch

File: src/libsystemd/sd-event/sd-event.c [hunk 2]
```diff
@@ -3893,6 +3906,11 @@ _public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
 
                                 assert(s);
 
+                                if (s->priority > threshold)
+                                        continue;
```
Comment (bluca) [inline src/libsystemd/sd-event/sd-event.c:3910]: this was a `break` in the previous patch

**Comment by bluca** [comment]
I've highlighted what (as far as I can see) are the main difference with the previous version the patch

**Comment by poettering** [comment]
still lgtm

**Comment by poettering** [comment]
@keszybz good to merge from your side too for 248?

**Comment by yuwata** [comment]
@bluca Thank you for listing the differences.

**Pull request merged**